src/crab/ebpf_domain.cpp

```diff
@@ -119,243 +121,674 @@ static std::vector<linear_constraint_t> assume_bit_cst_interval(const NumAbsDoma
+    auto lnub = left_interval_negative.truncate_to_sint(is64).ub();
+
     if (is64) {
-        // only in the 64bit case we can derive relational information
-        if (right_interval <= side_interval) {
-            using namespace crab::dsl_syntax;
-            if (is_lt) {
-                return {number_t{0} <= left, strict ? (left < right) : (left <= right)};
-            } else {
-                return {number_t{0} > left, strict ? (left > right) : (left >= right)};
-            }
-        }
+        if (!is_lt)
+            return assume_signed_64bit_gt(inv, strict, left_svalue, left_uvalue, left_interval_positive,
+                                          left_interval_negative, right_svalue, right_uvalue, right_interval);
+        else
+            return assume_signed_64bit_lt(inv, strict, left_svalue, left_uvalue, left_interval_positive,
+                                          left_interval_negative, right_svalue, right_uvalue, right_interval);
+    } else {
+        // 32-bit compare.
+        if (is_lt)
+            return assume_signed_32bit_lt(inv, strict, left_svalue, left_uvalue, left_interval_positive,
+                                          left_interval_negative, right_svalue, right_uvalue, right_interval);
+        else
+            return assume_signed_32bit_gt(inv, strict, left_svalue, left_uvalue, left_interval_positive,
+                                          left_interval_negative, right_svalue, right_uvalue, right_interval);
+    }
+    return {};
+}
 
-        if ((left_interval <= side_interval       && right_interval <= side_interval)
-         || (left_interval <= other_side_interval && right_interval <= other_side_interval)) {
-            // When no boundary is crossed, we can use the normal comparison.
-            using namespace crab::dsl_syntax;
-            if (is_lt) {
-                return {strict ? (left < right) : (left <= right)};
+static std::vector<linear_constraint_t> assume_unsigned_64bit_lt(const NumAbsDomain& inv, bool strict,
+                                                                 variable_t left_svalue, variable_t left_uvalue,
+                                                                 const crab::interval_t& left_interval_low,
+                                                                 const crab::interval_t& left_interval_high,
+                                                                 const linear_expression_t& right_svalue,
+                                                                 const linear_expression_t& right_uvalue,
+                                                                 const crab::interval_t& right_interval) {
+    using crab::interval_t;
+    using namespace crab::dsl_syntax;
+
+    auto rub = right_interval.ub();
+    auto lllb = left_interval_low.truncate_to_uint(true).lb();
+    auto llub = left_interval_low.truncate_to_uint(true).ub();
+    auto lhlb = left_interval_high.truncate_to_uint(true).lb();
```

Comment:
I think it's better to either rename or move the definition to immediately before use.

Reply:
Moved